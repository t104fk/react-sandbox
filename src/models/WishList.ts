import { ISimpleType, types } from 'mobx-state-tree';

type TWishListItem = {
  name: string,
  price: number,
  image?: string,
}
export const WishListItem = types.model({
  name: types.string,
  price: types.number,
  image: "",
}).actions(self => ({
  changeName(newName: string) {
    self.name = newName;
  },
  changePrice(newPrice: number) {
    self.price = newPrice;
  }
}))

export const WishList = types.model({
  items: types.optional(types.array(WishListItem), []),
})
.actions(self => ({
  add(item: TWishListItem) {
    self.items.push(item);
  }
}))
.views(self => ({
  get totalPrice() {
    return self.items.reduce((sum, entry) => sum + entry.price, 0);
  }
}));
