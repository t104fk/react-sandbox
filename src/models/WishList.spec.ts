import { getSnapshot, onPatch, onSnapshot }  from 'mobx-state-tree';
import { WishList, WishListItem } from "./WishList";

it("can create a instance of model", () => {
  const item = WishListItem.create({
    name: "Model name",
    price: 28.73,
    image:
      "https://www.talkwalker.com/images/2020/blog-headers/image-analysis.png",
  });
  expect(item.price).toBe(28.73);
  expect(item.image).toBe(
    "https://www.talkwalker.com/images/2020/blog-headers/image-analysis.png"
  );
  item.changeName('Natalie');
  expect(item.name).toBe('Natalie');
});

it("can create a wishlist", () => {
  const list = WishList.create({
    items: [
      {
        name: 'Model name',
        price: 28.73,
        image:
          "https://www.talkwalker.com/images/2020/blog-headers/image-analysis.png",
      },
    ],
  });
  expect(list.items.length).toBe(1);
  expect(list.items[0].name).toBe('Model name')
});

it('can add item', () => {
  const list = WishList.create();
  // list.add(WishListItem.create({
  //   name: 'Chesterton',
  //   price: 10,
  // }));

  // 状態遷移を保持するもの
  const states = [];

  // 新しいsnapshotの作成をlistenしているらしい
  // なので2回実行される
  // てことは対象のオブジェクトが変更されるたびsnapshotされる？
  // => オブジェクト変更されるたびstatesを更新してテストする
  onSnapshot(list, snapshot => {
    // listは追加後のもの(なのでname:Chesterton, length:1)
    // console.log(JSON.stringify(list), snapshot, states);
    // 状態遷移を保存しておく
    states.push(snapshot);
  })
  // createしなくてもmobxがやってくれるっぽい
  list.add({
    name: 'Chesterton',
    price: 10,
  });
  // console.log('add'); // 1回目のonSnapshotの後に実行されている => 1度目のsnapshot作成
  expect(list.items.length).toBe(1);
  expect(list.items[0].name).toBe('Chesterton');
  list.items[0].changeName('Chesterton jr.');
  console.log('changeName')
  expect(list.items[0].name).toBe('Chesterton jr.');

  // snapshotはcommitすべしとのこと
  // snapshotを作って、古いsnapshotと比較する
  // 失敗したらバグを修正もしくはsnapshotを更新する
  expect(getSnapshot(list)).toMatchSnapshot()

  // 2つ目のsnapshot testing
  expect(states).toMatchSnapshot();
})

it('can add item checking by patching', () => {
  const list = WishList.create();
  // 差分を保持する用
  const patches = [];
  onPatch(list, patch => {
    patches.push(patch);
  });

  // op: 'add'
  list.add({
    name: 'Chesterton',
    price: 10,
  });
  // op: 'replace'
  list.items[0].changeName('Chesterton jr.');

  // 差分の変更をsnapshotにしておき、それをテスト
  expect(patches).toMatchSnapshot();
})
