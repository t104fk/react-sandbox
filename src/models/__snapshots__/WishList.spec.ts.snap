// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add item 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "Chesterton jr.",
      "price": 10,
    },
  ],
}
`;

exports[`can add item 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Chesterton",
        "price": 10,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Chesterton jr.",
        "price": 10,
      },
    ],
  },
]
`;

exports[`can add item checking by patching 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "Chesterton",
      "price": 10,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "Chesterton jr.",
  },
]
`;
